Implement the feature of enrolling as a validator

linked0 opened the pull request:
Implmented basic methods for a Node to enroll as a validator.

- Add getUTXOs API endpoint
- Add createEnrollment API endpoint
- Add enrollValidator API endpoint
- Add hasEnrollment API endpoint

#406

TrustHenry commented on source/agora/node/Node.d:
```diff
@@ -338,9 +347,67 @@ public class Node : API
+    public UTXOSetValue[Hash] getUTXOs(TxType type, ulong min_amount_value) @safe
+    {
+        UTXOSetValue[Hash] result_utxos;
+        auto utxos = this.utxo_set.getUTXOs(config.node.key_pair.address);
```
This is to get an available UTXO list?, but not just the address of the node.
address must also be added by a parameter.

If it's for this node only, need to change the public.

TrustHenry commented on source/agora/node/API.d:
```diff
@@ -198,4 +200,73 @@ public interface API
+
+    ***************************************************************************/
+
+    public UTXOSetValue[Hash] getUTXOs(TxType type, ulong min_amount_value);
```
`public UTXOSetValue[Hash] getUTXOs (TxType type, ulong min_amount_value);`

TrustHenry replied:
```suggestion
    public UTXOSetValue[Hash] getUTXOs (TxType type, ulong min_amount_value);
```

TrustHenry commented on source/agora/node/Node.d:
```diff
@@ -338,9 +347,67 @@ public class Node : API
     }
+
+    /// GET: /get_utxos
+    public UTXOSetValue[Hash] getUTXOs(TxType type, ulong min_amount_value) @safe
```
```suggestion
    public UTXOSetValue[Hash] getUTXOs (TxType type, ulong min_amount_value) @safe
```

linked0 commented:
@Geod24 I have a CI build error on Linux as follows
```
Error: clean-up of Database incorrectly depends on destructors called by the GC.
Info: validator_set.dat��
./build/agora-unittests(_D4core7runtime18runModuleUnitTestsUZ19unittestSegvHandlerUNbiPSQCk3sys5posix6signal9siginfo_tPvZv+0x38)[0x5650f1ab7d0c]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7fe50ec6e890]
```
As you see the agora/tests/EnrollmentManager.d, All the code is commented out temporarily, but it outputs incomprehensible errors.
Does this remind you of anything?

As another option, I can close this PR and make multiple PRs.

Geod24 commented:
I think that, since you now instantiate a new EnrollmentManager for every node, some of them are not properly finalized (e.g. shutdown isn't being called).
You can probably test this by spinning up a VM and trying it.

linked0 commented:
Ready for review although the ci/circleci tests failed.
It's not obvious to me what failed [circle/ci result page](https://app.circleci.com/jobs/github/bpfkorea/agora/396)

Geod24 commented:
> source/agora/node/API.d(46,8): Error: module `Stellar_SCP` is in file 'scpd/types/Stellar_SCP.d' which cannot be read

You need to add it to [this array](https://github.com/bpfkorea/agora/blob/86ca56f29bcd4d2d580e2037b2e26ae622d42df6/tests/system/dub.json#L8). We should really fix this.

linked0 commented:
The collect2 outputs error on the circle/ci test as follows after import paths being added.

> collect2: error: ld returned 1 exit status

Geod24 commented:
Looking at the link error, you need to add the sqlite dependency in [here](https://github.com/bpfkorea/agora/blob/86ca56f29bcd4d2d580e2037b2e26ae622d42df6/tests/system/dub.json#L38-L44)

linked0 commented:
In order to fix the ci/circleci test errors, I have modified `tests/system/dub.json` file as follows.
1. add import paths to the "importPaths" section.
2. add references to the "sourceFiles" section.
3. add references for submodules to the "dependencies" section.

linked0 commented:
Ready for review.

AndrejMitrovic commented on source/agora/node/API.d:
```diff
@@ -198,4 +200,75 @@ public interface API
+
+    ***************************************************************************/
+
+    public Enrollment createEnrollment (Hash frozen_utxo_hash);
```
And if the utxo doesn't exist, what happens then?

linked0 replied:
Your question reminds me that I should check the validity of the enrollment data in the receiveEnrollment function of the GossipProtocol class and the validity of the UTXO in the createEnrollment function that you mentioned above.

linked0 replied:
Herny is making the function for checking validity of UTXO called isInvalidValidatorUTXO. So I will use it after his work is done.

AndrejMitrovic commented:
Looks good, but I'm not sure about introducing several different APIs in the same commit. Maybe it makes sense if they cannot exist without being grouped together like that. I'll let others share their opinions.

linked0 commented:
> Looks good, but I'm not sure about introducing several different APIs in the same commit. Maybe it makes sense if they cannot exist without being grouped together like that. I'll let others share their opinions.

The APIs must have been grouped together because I had to test the entire process from start to end of the `enrollValidator` API. There are some difficulties in unit testing and fixing bugs because if there are some errors at the commit few steps before, a developer must rebase all the commits. It must be unnatural and inconvenient for developing.

Anyway, I'll make sure to pay more careful attention to what you said while working

linked0 commented:
Ready for review.
For Drey's comment, I will make another issue for validation check for enrollment data.

TrustHenry reviewed (approved):
LGTM

AndrejMitrovic commented on source/agora/node/API.d:
```diff
@@ -198,4 +200,75 @@ public interface API
+    ***************************************************************************/
+
+    @method(HTTPMethod.GET)
+    public bool hasEnrollment (Hash enroll_hash);
```
What is the use-case for this API? I see that it's used in the unittest, but other than the tests why do we need to query another node if it contains an enrollment hash?

linked0 replied:
I follow the way the code about transaction has implemented. It has the hasTransactionHash function so I add the API similar to that. Its usage is just for the unit test now.

AndrejMitrovic replied:
But in that case, `NetworkClient.sendTransaction` calls `hasTransactionHash` before calling `putTransaction`, to avoid sending the same transaction to a node which already contains it. That was its use-case.

Then the same should be done for the enrollment, right? `NetworkManager.sendEnrollment` should first check `hasEnrollment` before calling `enrollValidator` to avoid spamming the network.

linked0 replied:
That's right. I have missed it. Thanks.

linked0 replied:
> Then the same should be done for the enrollment, right? 
> NetworkManager.sendEnrollment should first check hasEnrollment before calling enrollValidator to avoid spamming the network.

Please ignore what I commented before(=> "That's right. I have missed it.")
It is already checked in receiveEnrollment function of the GossipProtocol class, actually in the addEnrollment function in the receiveEnrollment function. The sendTransaction you mentioned is not checking other node's state also.

AndrejMitrovic replied:
But if it's checked in `receiveEnrollment` it means the structure was already sent over the wire. The point is to not send data that doesn't have to be sent.

linked0 replied:
I got it. I will be checking the code more.

linked0 replied:
I will add the code calling the hasEnrollment function into the sendEnrollement function of the **NetworkClient** class like Transaction.

AndrejMitrovic commented on source/agora/node/Node.d:
```diff
@@ -338,9 +347,70 @@ public class Node : API
     }
+
+    /// GET: /get_utxos
+    public UTXOSetValue[Hash] getUTXOs (PublicKey address, TxType type,
```
This method is allocating two associative arrays. We could add an optional `min_amount` parameter to  `UTXOSet.getUTXOs` and implement the amount filtering there?

linked0 replied:
Sure, we could. The function will be enhanced in another PR.

linked0 commented:
Ready for review
1) Make use the hasEnrollment function in the gossip protocol.
2) Resolve an synchronization issue between nodes.(See agora/tests/EnrollmentManager.d)

linked0 commented:
Rebased.

Geod24 commented on source/agora/node/API.d:
```diff
@@ -198,4 +200,75 @@ public interface API
+
+    ***************************************************************************/
+
+    public UTXOSetValue[Hash] getUTXOs (PublicKey address, TxType type,
```
Why is this in the API ?? Same question for `createEnrollment` ?

linked0 replied:
The getUTXOs can be used in the situation when a user(or Node client app) select an UTXO as a deposit for enrollment. So it should be open as API.
But, the createEnrollment could be inserted in the enrollValidator which must take an UTXO as parameter.

linked0 replied:
`enrollVaidator` is an API endpoint receiving the enrollment data and cannot make enrollment data, and an user client should select an UTXO in the current scenario. So createEnrollment is needed now as a bridge between the two APIs. After the operations of an user Node client is determined, the createEnrollment API could be deleted.

Geod24 replied:
> The getUTXOs can be used in the situation when a user(or Node client app) select an UTXO as a deposit for enrollment. So it shoud be open as API.

Everything in `API` should be used *only* for interfacing with other nodes. There should not be *any* method here for convenience (by convenience, I mean used by non-node, e.g. APIs for exchanges).
Because such methods are easy DoS endpoint. Imagine a few clients calling this method if you have even [1% of the amount of UTXOs Bitcoin has today](https://www.blockchain.com/en/charts/utxo-count).

For convenience methods, we would expose *another* API, which is only enabled on demand. An API expected to be used only locally does not need to worry about this.

Geod24 replied:
> `enrollVaidator` is an API endpoint receiving the enrollment data and cannot make enrollment data, and an user client should select an UTXO in the current scenario. So `createEnrollment` is needed now as a bridge between the two APIs.

If you want to force a validator to enroll, just make an `enroll` method in [here](https://github.com/bpfkorea/agora/blob/1010eb23fdc9d702f5eea3e6dc4f30027da3dd73/source/agora/test/Base.d#L401-L415). This will not expose it to the world, but allow you to trigger specific actions on a node from an unittest.

Geod24 commented on source/agora/node/API.d:
```diff
@@ -198,4 +200,75 @@ public interface API
+    ***************************************************************************/
+
+    public UTXOSetValue[Hash] getUTXOs (PublicKey address, TxType type,
+        ulong min_amount_value);
```
Any amount should be expressed using `Amount`

Geod24 commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -0,0 +1,97 @@
+    import core.thread;
+
+    const NodeCount = 4;
+    const SpendAmount = 40_000 * 10_000_000;
```
Use `Amount` in order to avoid to have to do this

Geod24 commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -0,0 +1,97 @@
+        auto has_enroll = node_2.hasEnrollment(utxo_hashes[0]);
+        // Retry if we're too early
+        if (has_enroll == false && times++ < 50)
+            goto Start;
```
You don't need a goto here. The code I showed you as an example needed it because there was something after in the loop, but it's not the case here.
E.g.:
```D
while (!node_2.hasEnrollment(...) && times++ < 50)
    Thread.sleep(10.msecs);
assert(times < 50);
```

This will assert if it reaches the 500 msecs timeout

AndrejMitrovic replied:
Why not just use the `retryFor` function like the rest of the test-suite?

Geod24 replied:
![Reasons](https://media.giphy.com/media/x43pXtJShv93a/source.gif)

Geod24 commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -0,0 +1,78 @@
+    import std.algorithm;
+    import std.conv;
+    import core.time;
+    import core.thread;
```
The `core.thread` import is not necessary. The other ones should be moved to global scope.

Geod24 commented on source/agora/node/Node.d:
```diff
@@ -387,9 +397,81 @@ public class Node : API
+
+    ***************************************************************************/
+
+    protected UTXOSetValue[Hash] getUTXOs (PublicKey address) @safe
```
Since this function is only forwarding, and it's protected, you can remove it.
Instead call `this.utxo_set.getUTXOs` directly.

But I realize that most members of `Node` are `private`. Perhaps they should be made `protected`. Could you do this in another commit / PR (since it's a no brainer, another PR would get merged very quickly).

linked0 replied:
> Could you do this in another commit / PR (since it's a no brainer, another PR would get merged very quickly).

Sure, I will do that in another commit of this PR.

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -125,6 +136,16 @@ public class EnrollmentManager
+        // check validity of the enrollment data
+        auto reason = isInvalidEnrollmentReason(this.ledger.getBlockHeight() + 1,
+            enroll, this.utxo_set.getUTXOFinder());
+        if (reason !is null)
```
One nice idiom in D is the ability to do:
```D
if (auto reason = func())
```

Which looks like a perfect fit for this.

Geod24 reviewed (approved):
Looks very good, just a couple suggestions and we're good to go

linked0 commented:
Ready for review. 
Is it OK to reqeust a review in spite of failing in all tests on Github/Linux? If any problem, please tell me.

AndrejMitrovic commented on .github/workflows/github.yml:
```diff
@@ -25,6 +25,8 @@ jobs:
     - name: Build & test agora
       run: |
+        ulimit -a
+        ulimit -Ss 16384
```
These two should be swapped. So we can verify the limits were increased.

AndrejMitrovic commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -125,6 +136,15 @@ public class EnrollmentManager
+        if (auto reason = isInvalidEnrollmentReason(this.ledger.getBlockHeight() + 1,
+            enroll, this.utxo_set.getUTXOFinder()))
+        {
+            this.logMessage("Invalid enrollment data, Reason: " ~ reason,
```
This will allocate a new string every time enrollment fails validation. See this comment here: https://github.com/bpfkorea/agora/pull/465#discussion_r364074502

linked0 replied:
> This will allocate a new string every time enrollment fails validation. See this comment here: #465 (comment)

The logMessage function should be enhanced for general usage also. In the PR for the enhancement, I will change the code with your comment.

linked0 committed:
Make members of Node protected

The TestNode class needs to access members of the Node class

linked0 commented:
The log of the Github/Linux test error is like as follows.
```
Running ./build/agora-unittests 
./build/agora-unittests(_D4core7runtime18runModuleUnitTestsUZ19unittestSegvHandlerUNbiPSQCk3sys5posix6signal9siginfo_tPvZv+0x38)[0x5586650364e8]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7f30f1789890]
./build/agora-unittests(_D5agora9consensus4data7UTXOSetQi8findUTXOMFNbNfS6geod247bitblob__T7BitBlobVmi512ZQpmJSQDhQDeQCxQCv12UTXOSetValueZb+0xf9)[0x558664aba529]
./build/agora-unittests(_D5agora9consensus10Validation25isInvalidEnrollmentReasonFNbNfxmKxSQCnQCk4data10EnrollmentQmDFNbNfS6geod247bitblob__T7BitBlobVmi512ZQpmJSQFfQFcQCs7UTXOSet12UTXOSetValueZbZAya+0x6f)[0x558664ab076b]
./build/agora-unittests(_D5agora6common17EnrollmentManagerQt13addEnrollmentMFNbNfKxSQCg9consensus4data10EnrollmentQmZb+0x89)[0x558664a19aa1]
./build/agora-unittests(_D5agora4node14GossipProtocolQq17receiveEnrollmentMFNfSQCb9consensus4data10EnrollmentQmZv+0x29)[0x558664afa1b5]
./build/agora-unittests(_D5agora4node4NodeQf15enrollValidatorMFNfSQBo9consensus4data10EnrollmentQmZv+0xd5)[0x558664b02849]
./build/agora-unittests(_D6geod249LocalRest__T9RemoteAPITC5agora4test4Base7TestAPIZQBl13handleCommandFSQCzQCv7CommandQCiSQDrQDn9FilterAPIZv+0x1b31)[0x558664b079a9]
```

linked0 commented:
Another error log on the Github/Linux tests
```
Running ./build/agora-unittests 
./build/agora-unittests(_D4core7runtime18runModuleUnitTestsUZ19unittestSegvHandlerUNbiPSQCk3sys5posix6signal9siginfo_tPvZv+0x38)[0x56003afa64e8]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7f266b68a890]
./build/agora-unittests(_D5agora4node4NodeQf13hasEnrollmentMFNfS6geod247bitblob__T7BitBlobVmi512ZQpZb+0x21)[0x56003aa7287d]
./build/agora-unittests(_D6geod249LocalRest__T9RemoteAPITC5agora4test4Base7TestAPIZQBl13handleCommandFSQCzQCv7CommandQCiSQDrQDn9FilterAPIZv+0x1d4f)[0x56003aa77bc7]
./build/agora-unittests(_D6geod249LocalRest__T9RemoteAPITC5agora4test4Base7TestAPIZQBl__T7spawnedTCQBpQBmQBk8TestNodeZQBdFSQCn6common6ConfigQhPOSQEp8RegistryQjZ__T6handleTSQFqQFm7CommandZQyMFQuZ9__lambda2MFZv+0x52)[0x56003aa80bc2]
./build/agora-unittests(_D6geod2411concurrency14FiberScheduler6createMFNbDFZvZ4wrapMFZv+0x2b)[0x56003ac85297]
./build/agora-unittests(_D4core6thread5fiber5Fiber3runMFZv+0x2a)[0x56003afa72ae]
./build/agora-unittests(fiber_entryPoint+0x63)[0x56003afa7193]
Program exited with code -11
```

linked0 committed:
Implement the feature of enrolling as a validator

- Add enrollValidator API endpoint
- Add hasEnrollment API endpoint
- unittest code

linked0 commented:
Ready for review. 
- I added the code checking validity of an enrollment data
- I resoved the unittest error.

**Unittest Error Description**
After the flow of the unittest code in the agora.test.EnrollmentManager module ended, the following code in the sendEnrollment function of NetworkClient was running in some thread and crashed with the Segfault because the network including all nodes was shutdowned.
```
this.taskman.runTask(
{
    // if the node already has this enrollment, don't send it
    if (this.attemptRequest!(LogLevel.Trace)(
        this.api.hasEnrollment(enroll.utxo_key), null))
            return;
    this.attemptRequest(this.api.enrollValidator(enroll), null);
});
```

**Techniques used to analyze this issue**
- dub test --compiler=ldc2
- lldb ./build/agora-unittests
```
(lldb) run
(lldb) p this
(lldb) p this.enroll_man
```
- `shared` variable for selective log print
- checking the value of pointer with the following snippet
```
void** ffs = cast(void**)&this.enroll_man;
*ffs = cast(void*)0xDEAD_BEEF;
```
- and Henry's PR #375
